src/json-patch-duplex.js

```diff
@@ -518,10 +521,15 @@ var jsonpatch;
                         if (validateOperation && operation.op === "add" && key > obj.length) {
                             throw new JsonPatchError("The specified index MUST NOT be greater than the number of elements in the array", "OPERATION_VALUE_OUT_OF_BOUNDS", 0, operation.path, operation);
                         }
-                        returnValue.result = arrOps[operation.op].call(operation, obj, key, document); // Apply patch
-                        returnValue.newDocument = document;
-                        if (returnValue.result === false) {
-                            throw new JsonPatchError("Test operation failed", 'TEST_OPERATION_FAILED', 0, operation, document);
+                        if (operation.op === 'test') {
+                            returnValue.test = arrOps[operation.op].call(operation, obj, key, document); // Apply patch
+                            if (returnValue.test === false) {
+                                throw new JsonPatchError("Test operation failed", 'TEST_OPERATION_FAILED', 0, operation, document);
+                            }
+                        }
+                        else {
+                            debugger;
+                            returnValue.removed = arrOps[operation.op].call(operation, obj, key, document); // Apply patch
```

Comment:
I think it would be cleaner if each `arrOps[*]` would return applicable object `{remove: smth}` or `{test: true}` so you will not have to check it once again here.